teos/src/api/tor.rs

```diff
@@ -0,0 +1,106 @@
+use anyhow::{bail, Result};
+use std::net::{IpAddr, Ipv4Addr, SocketAddr, SocketAddrV4};
+use tokio::net::TcpStream;
+use tokio::time::{sleep, Duration};
+use torut::control::UnauthenticatedConn;
+use torut::onion::TorSecretKeyV3;
+
+//
+// Expose an onion service that re-directs to the public api.
+//
+pub async fn expose_onion_service(
+    tor_control_port: u16,
+    api_port: u16,
+    onion_port: u16,
+) -> Result<()> {
+    let stream = match connect_tor_cp(SocketAddr::V4(SocketAddrV4::new(
+        Ipv4Addr::LOCALHOST,
+        tor_control_port,
+    )))
+    .await
+    {
+        Ok(info) => info,
+        Err(err) => bail!(err),
+    };
+
+    let mut utc = UnauthenticatedConn::new(stream);
+
+    let info = match utc.load_protocol_info().await {
+        Ok(info) => info,
+        Err(error) => bail!("failed to load protocol info from Tor: {}", error),
+    };
+    let ad = info.make_auth_data()?.expect("failed to make auth data");
+
+    if utc.authenticate(&ad).await.is_err() {
+        bail!("failed to authenticate with Tor")
+    }
+
+    let mut ac = utc.into_authenticated().await;
+
+    ac.set_async_event_handler(Some(|_| async move { Ok(()) }));
```

Comment:
Sadly no, the set_async_event_handler needs a function passed into it so when an asynchronous event happens AuthenticatedConn can handle the event, the function that handles the event is required to be set, from the docs in the source file of the libraries function:
`
 # Async event handling
/// AuthenticatedConn automatically recognises and treats differently response for given request and asynchronous
/// event response.
/// If it receives an asynchronous event it will invoke async event handler(if some).
/// Asynchronous handler will be awaited in current thread(no calls to `tokio::spawn` or stuff like that).
/// Make sure that async handlers won't take long time to execute as this may cause latencies in handling other functions.
///
/// Please also note that this connection won't do anything in background to handle events.
/// In order to trigger event handling(if any) use `noop` function.
`

And when the parameter passed into set_async_event_handler is None the compiler is unable to figure out what to set F as in this

`
      impl<S, H, F> AuthenticatedConn<S, H>
      where
          S: AsyncRead + Unpin,
      // there fns make use of event handler so it's needed
          H: Fn(AsyncEvent<'static>) -> F,
          F: Future<Output=Result<(), ConnError>>,
  {
   
`